File: packages/protocol/src/session-control.ts
// Slice 7: typed mirror of schemas/session-control-v1.json (the session
// CONTROL plane: WebSocket JSON, protocolVersion = 1). The JSON file is the
// single source of truth; session-control.test.ts pins this mirror to it.
// Data plane (binary mesh) stays in schemas/cad_protocol.fbs + decodeMesh*.

import { z } from "zod";

export const SESSION_CONTROL_VERSION = 1 as const;

/** Relay control methods (flat replies). */
export const CONTROL_METHODS = [
  "hello",
  "snapshot",
  "invoke",
  "txnBegin",
  "txnCommit",
  "txnRollback",
  "txnForceRollback",
  "txnStatus",
] as const;
export type ControlMethod = (typeof CONTROL_METHODS)[number];

/** Read-only query methods ({result} replies; txn-adjacent previews included). */
export const QUERY_METHODS_CONTRACT = [
  "getDocumentInfo",
  "getBodies",
  "getFeatures",
  "getFeature",
  "getParameters",
  "getDependencies",
  "getModelTree",
  "describeModel",
  "getSelection",
  "setSelection",
  "clearSelection",
  "findFaces",
  "findEdges",
  "findBodies",
  "getManipulators",
  "measureVolume",
  "measureArea",
  "getBoundingBox",
  "measureDistance",
  "measureAngle",
  "measureRadius",
  "measureDiameter",
  "validateDocument",
  "validateBody",
  "validateFeature",
  "listCommands",
  "getCommandSchema",
  "getCapabilities",
  "previewBegin",
  "previewUpdate",
  "previewCommit",
  "previewCancel",
] as const;
export type QueryMethodContract = (typeof QUERY_METHODS_CONTRACT)[number];

/** Every control-plane method, in contract order. */
export const SESSION_CONTROL_METHODS: readonly string[] = [
  ...CONTROL_METHODS,
  ...QUERY_METHODS_CONTRACT,
];

/** Relay/query-enforced required params per method (missing → coded error). */
export const REQUIRED_PARAMS: Record<string, readonly string[]> = {
  hello: ["token", "protocolVersion"],
  snapshot: [],
  invoke: ["type"],
  txnBegin: ["transactionId"],
  txnCommit: ["transactionId"],
  txnRollback: ["transactionId"],
  txnForceRollback: ["transactionId"],
  txnStatus: [],
  getDocumentInfo: [],
  getBodies: [],
  getFeatures: [],
  getFeature: ["featureId"],
  getParameters: ["featureId"],
  getDependencies: ["featureId"],
  getModelTree: [],
  describeModel: [],
  getSelection: [],
  setSelection: ["ids"],
  clearSelection: [],
  findFaces: [],
  findEdges: [],
  findBodies: [],
  getManipulators: ["featureId"],
  measureVolume: [],
  measureArea: ["featureId"],
  getBoundingBox: ["featureId"],
  measureDistance: ["a", "b"],
  measureAngle: ["a", "b"],
  measureRadius: ["featureId"],
  measureDiameter: ["featureId"],
  validateDocument: [],
  validateBody: [],
  validateFeature: [],
  listCommands: [],
  getCommandSchema: [],
  getCapabilities: [],
  previewBegin: ["featureId", "paramName"],
  previewUpdate: ["previewId"],
  previewCommit: ["previewId"],
  previewCancel: ["previewId"],
};

export const SERVER_EVENTS = ["delta", "selection", "core-restarted"] as const;

// ── Typed DTOs for the fixed control messages (top 8; open-ended query
// params stay Record<string, unknown> — follow-up, not half-migrated). ──

export const HelloParamsSchema = z.object({
  token: z.string().min(1),
  protocolVersion: z.literal(SESSION_CONTROL_VERSION),
  clientType: z.string().optional(),
  clientName: z.string().optional(),
});
export type HelloParams = z.infer<typeof HelloParamsSchema>;

export const SnapshotParamsSchema = z.object({
  documentId: z.string().min(1).optional(),
});
export type SnapshotParams = z.infer<typeof SnapshotParamsSchema>;

export const InvokeParamsSchema = z.object({
  documentId: z.string().min(1).optional(),
  type: z.number().int(),
  fields: z.record(z.unknown()).optional(),
  baseRevision: z.number().nullable().optional(),
  transactionId: z.string().min(1).optional(),
});
export type InvokeParams = z.infer<typeof InvokeParamsSchema>;

export const TxnParamsSchema = z.object({
  documentId: z.string().min(1).optional(),
  transactionId: z.string().min(1),
});
export type TxnParams = z.infer<typeof TxnParamsSchema>;

export const ErrorReplySchema = z.object({
  requestId: z.string().min(1),
  ok: z.literal(false),
  errorCode: z.string().min(1),
  error: z.string().min(1),
});
export type ErrorReply = z.infer<typeof ErrorReplySchema>;

File: clients/session-dotnet/src/Kreoda.SessionClient/SessionControl.cs
// Slice 7: typed mirror of schemas/session-control-v1.json (session CONTROL
// plane: WebSocket JSON, protocolVersion = 1). The JSON file is the single
// source of truth; SessionControlContractTests pins this mirror to it.
// The wire stays Dictionary-built JSON inside one place (ToDictionary);
// open-ended query params keep using CallAsync directly (follow-up).
namespace Kreoda.Session;

/// <summary>Control-plane method names (contract order).</summary>
public static class SessionMethods
{
    public const string Hello = "hello";
    public const string Snapshot = "snapshot";
    public const string Invoke = "invoke";
    public const string TxnBegin = "txnBegin";
    public const string TxnCommit = "txnCommit";
    public const string TxnRollback = "txnRollback";
    public const string TxnForceRollback = "txnForceRollback";
    public const string TxnStatus = "txnStatus";

    public static readonly string[] All =
    [
        "hello", "snapshot", "invoke",
        "txnBegin", "txnCommit", "txnRollback", "txnForceRollback", "txnStatus",
        "getDocumentInfo", "getBodies", "getFeatures", "getFeature",
        "getParameters", "getDependencies", "getModelTree", "describeModel",
        "getSelection", "setSelection", "clearSelection",
        "findFaces", "findEdges", "findBodies",
        "getManipulators", "measureVolume", "measureArea", "getBoundingBox",
        "measureDistance", "measureAngle", "measureRadius", "measureDiameter",
        "validateDocument", "validateBody", "validateFeature",
        "listCommands", "getCommandSchema", "getCapabilities",
        "previewBegin", "previewUpdate", "previewCommit", "previewCancel",
    ];

    /// <summary>Relay/query-enforced required params per method.</summary>
    public static readonly IReadOnlyDictionary<string, string[]> RequiredParams =
        new Dictionary<string, string[]>
        {
            ["hello"] = ["token", "protocolVersion"],
            ["snapshot"] = [],
            ["invoke"] = ["type"],
            ["txnBegin"] = ["transactionId"],
            ["txnCommit"] = ["transactionId"],
            ["txnRollback"] = ["transactionId"],
            ["txnForceRollback"] = ["transactionId"],
            ["txnStatus"] = [],
            ["getDocumentInfo"] = [],
            ["getBodies"] = [],
            ["getFeatures"] = [],
            ["getFeature"] = ["featureId"],
            ["getParameters"] = ["featureId"],
            ["getDependencies"] = ["featureId"],
            ["getModelTree"] = [],
            ["describeModel"] = [],
            ["getSelection"] = [],
            ["setSelection"] = ["ids"],
            ["clearSelection"] = [],
            ["findFaces"] = [],
            ["findEdges"] = [],
            ["findBodies"] = [],
            ["getManipulators"] = ["featureId"],
            ["measureVolume"] = [],
            ["measureArea"] = ["featureId"],
            ["getBoundingBox"] = ["featureId"],
            ["measureDistance"] = ["a", "b"],
            ["measureAngle"] = ["a", "b"],
            ["measureRadius"] = ["featureId"],
            ["measureDiameter"] = ["featureId"],
            ["validateDocument"] = [],
            ["validateBody"] = [],
            ["validateFeature"] = [],
            ["listCommands"] = [],
            ["getCommandSchema"] = [],
            ["getCapabilities"] = [],
            ["previewBegin"] = ["featureId", "paramName"],
            ["previewUpdate"] = ["previewId"],
            ["previewCommit"] = ["previewId"],
            ["previewCancel"] = ["previewId"],
        };
}

/// <summary>Typed <c>invoke</c> params (fields stay open: core command payloads).</summary>
public sealed record InvokeRequest(
    int Type,
    IDictionary<string, object?> Fields,
    string DocumentId = "doc-phase1",
    double? BaseRevision = null,
    string? TransactionId = null)
{
    public Dictionary<string, object?> ToDictionary()
    {
        var p = new Dictionary<string, object?>
        {
            ["documentId"] = DocumentId,
            ["type"] = Type,
            ["fields"] = Fields,
        };
        if (BaseRevision.HasValue) p["baseRevision"] = BaseRevision.Value;
        if (TransactionId is not null) p["transactionId"] = TransactionId;
        return p;
    }
}

/// <summary>Typed transaction-control params.</summary>
public sealed record TxnRequest(string TransactionId, string DocumentId = "doc-phase1")
{
    public Dictionary<string, object?> ToDictionary() => new()
    {
        ["documentId"] = DocumentId,
        ["transactionId"] = TransactionId,
    };
}
